File: src/cmds/typescript.ts
import { run } from '../core/runner';
import { stripAnsi } from '../core/utils';
import { HandlerResult } from './git';

export function handleTsc(args: string[], ultra: boolean): HandlerResult {
  const result = run('tsc', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';

  try {
    filtered = groupErrors(raw, ultra);
  } catch {
    filtered = raw;
  }

  return {
    filteredOutput: filtered || (result.exitCode === 0 ? (ultra ? '✓' : '✓ no errors') : raw),
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType: 'tsc',
    execMs: result.execMs,
  };
}

function groupErrors(raw: string, ultra = false): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n');

  type ErrInfo = { msg: string; count: number };
  const byFile = new Map<string, Map<string, ErrInfo>>();
  let totalErrors = 0;

  for (const line of lines) {
    const m = /^(.+?)\((\d+),(\d+)\):\s+error\s+(TS\d+):\s+(.+)/.exec(line);
    if (!m) continue;
    const file = m[1];
    const code = m[4];
    const msg = m[5];
    totalErrors++;
    let fileMap = byFile.get(file);
    if (!fileMap) {
      fileMap = new Map();
      byFile.set(file, fileMap);
    }
    const existing = fileMap.get(code);
    if (existing) existing.count++;
    else fileMap.set(code, { msg, count: 1 });
  }

  if (totalErrors === 0) {
    return ultra ? '✓' : '✓ no errors';
  }

  if (ultra) {
    const parts: string[] = [];
    for (const [file, codes] of byFile) {
      const fileShort = file.split(/[\/\\]/).pop() || file;
      const codeStrs: string[] = [];
      for (const [code, info] of codes) {
        codeStrs.push(`${code}×${info.count}`);
      }
      parts.push(`${fileShort}:${codeStrs.join(',')}`);
    }
    return `${totalErrors}E/${byFile.size}F: ${parts.slice(0, 5).join(' ')}`;
  }

  const out: string[] = [];
  for (const [file, codes] of byFile) {
    const total = Array.from(codes.values()).reduce((s, e) => s + e.count, 0);
    out.push(`${file}: ${total} error${total === 1 ? '' : 's'}`);
    for (const [code, info] of codes) {
      const xN = info.count > 1 ? `(×${info.count}) ` : `(×${info.count}) `;
      out.push(`  ${code} ${xN}${info.msg}`);
    }
  }
  out.push('');
  out.push(`Total: ${totalErrors} error${totalErrors === 1 ? '' : 's'} in ${byFile.size} file${byFile.size === 1 ? '' : 's'}`);
  return out.join('\n');
}

File: internal/handlers/typescript.go
package handlers

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/Kevalgor12/tok-proxy/internal/run"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

func HandleTsc(args []string, ultra bool) Result {
	r := run.Run("tsc", args)
	raw := combined(r)
	filtered := groupTscErrors(raw, ultra)

	if filtered == "" {
		if r.ExitCode == 0 {
			if ultra {
				filtered = "✓"
			} else {
				filtered = "✓ no errors"
			}
		} else {
			filtered = raw
		}
	}
	return Result{Filtered: filtered, Exit: r.ExitCode, Raw: raw, CmdType: "tsc", ExecMs: r.ExecMs}
}

var tscErrLineRe = regexp.MustCompile(`^(.+?)\((\d+),(\d+)\):\s+error\s+(TS\d+):\s+(.+)`)

// groupTscErrors folds "file(line,col): error TSxxxx: msg" lines into per-file, per-code
// counts, preserving first-seen order (Go maps don't, so files/codes are tracked in slices).
func groupTscErrors(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)

	type codeInfo struct {
		msg   string
		count int
	}
	type fileErrs struct {
		codes  []string
		byCode map[string]*codeInfo
	}
	var fileOrder []string
	files := map[string]*fileErrs{}
	total := 0

	for _, line := range strings.Split(clean, "\n") {
		m := tscErrLineRe.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		file, code, msg := m[1], m[4], m[5]
		total++
		fe, ok := files[file]
		if !ok {
			fe = &fileErrs{byCode: map[string]*codeInfo{}}
			files[file] = fe
			fileOrder = append(fileOrder, file)
		}
		if info, ok := fe.byCode[code]; ok {
			info.count++
		} else {
			fe.byCode[code] = &codeInfo{msg: msg, count: 1}
			fe.codes = append(fe.codes, code)
		}
	}

	if total == 0 {
		if ultra {
			return "✓"
		}
		return "✓ no errors"
	}

	if ultra {
		var parts []string
		for _, file := range fileOrder {
			fe := files[file]
			short := file
			if i := strings.LastIndexAny(file, `/\`); i >= 0 {
				short = file[i+1:]
			}
			var codeStrs []string
			for _, code := range fe.codes {
				codeStrs = append(codeStrs, fmt.Sprintf("%s×%d", code, fe.byCode[code].count))
			}
			parts = append(parts, short+":"+strings.Join(codeStrs, ","))
		}
		return fmt.Sprintf("%dE/%dF: %s", total, len(fileOrder), strings.Join(limit(parts, 5), " "))
	}

	var out []string
	for _, file := range fileOrder {
		fe := files[file]
		fileTotal := 0
		for _, code := range fe.codes {
			fileTotal += fe.byCode[code].count
		}
		out = append(out, fmt.Sprintf("%s: %d %s", file, fileTotal, plural(fileTotal, "error", "errors")))
		for _, code := range fe.codes {
			info := fe.byCode[code]
			out = append(out, fmt.Sprintf("  %s (×%d) %s", code, info.count, info.msg))
		}
	}
	out = append(out, "", fmt.Sprintf("Total: %d %s in %d %s",
		total, plural(total, "error", "errors"), len(fileOrder), plural(len(fileOrder), "file", "files")))
	return strings.Join(out, "\n")
}
